libraries/utils/generics/dyncircularbuffer.h

```diff
@@ -0,0 +1,521 @@
+* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+* POSSIBILITY OF SUCH DAMAGE.
+*
+*
+* @brief     DynamicCircularBuffer class declaration.
+*
+*/
+
+#ifndef IOBUFFER_DYNAMICCIRCULARBUFFER_H
+#define IOBUFFER_DYNAMICCIRCULARBUFFER_H
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// INCLUDES
+//=============================================================================================================
+
+#include "../utils_global.h"
+
+//*************************************************************************************************************
+//=============================================================================================================
+// QT INCLUDES
+//=============================================================================================================
+
+#include <QSharedPointer>
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// Eigen INCLUDES
+//=============================================================================================================
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// FORWARD DECLARATIONS
+//=============================================================================================================
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// DEFINE NAMESPACE IOBUFFER
+//=============================================================================================================
+
+namespace IOBUFFER {
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// IOBUFFER FORWARD DECLARATIONS
+//=============================================================================================================
+
+
+//=============================================================================================================
+/**
+* This class is a implementation of a queue based on a array with dynamic size.
+*
+* @brief Dynamic circular queue/buffer.
+*/
+template<typename T>
+class DynCircularBuffer
```

Comment:
Dynamic circular buffer is not really useful. you could use a queue instead

Reply:
It is a queue, but array-based for faster access and the [ ] operator can be used. You are right i should rename it.

Reply:
And why has it to be circular than?

Reply:
It minimises the number of  memory allocations, because space that is set free by pop_front() can later be used by push_back(). Maybe this is helpful:
![arrayqueue](https://user-images.githubusercontent.com/23101388/30980241-a38790c2-a480-11e7-8e64-f2d6ff53c0dc.png)

Reply:
I see, you basically safe the request and allocation of new memory from the heap.

Reply:
Yes allocation is only needed when all slots in the array are full, when this happens the size is doubled.

Reply:
Yesterday before i answerd your questions @LorenzE told me that you guys would prefer QLinkedList for the workers, is this still true? Please let me know.

Reply:
@1DIce: yes please use the QLinkedList